src/cdomains/arrayDomain.ml

```diff
@@ -941,7 +994,746 @@ struct
+    | Some i -> (try Some (Z.of_int (Z.to_int i)) with Z.Overflow -> None)
+    | None -> None
+
+  let get ?(checkBounds=true) (ask: VDQ.t) (must_nulls_set, may_nulls_set, size) (e, i) =
+    let rec all_indexes_must_null i max =
+      if Z.gt i max then
+        true
+      else if MustNulls.mem i must_nulls_set then
+        all_indexes_must_null (Z.succ i) max
+      else
+        false in
+    let min interval = match Idx.minimal interval with
+      | Some min_num when Z.geq min_num Z.zero -> min_num
+      | _ -> Z.zero in (* assume worst case minimal natural number *)
+
+    let min_i = min i in
+    let max_i = idx_maximal i in
+    let min_size = min size in
+
+    (* warn if index is (potentially) out of bounds *)
+    if checkBounds then (array_oob_check (module Idx) ((must_nulls_set, may_nulls_set), size) (e, i));
+    match max_i, idx_maximal size with
+    (* if there is no maximum value in index interval *)
+    | None, _ ->
+      (* ... return NotNull if no i >= min_i in may_nulls_set *)
+      if not (MayNulls.exists (Z.leq min_i) may_nulls_set) then
+        NotNull
+        (* ... else return Top *)
+      else
+        Top
+    (* if there is no maximum size *)
+    | Some max_i, None when Z.geq max_i Z.zero ->
+      (* ... and maximum value in index interval < minimal size, return Null if all numbers in index interval are in must_nulls_set *)
+      if Z.lt max_i min_size && all_indexes_must_null min_i max_i then
+        Null
+        (* ... return NotNull if no number in index interval is in may_nulls_set *)
+      else if not (MayNulls.exists (fun x -> Z.geq x min_i && Z.leq x max_i) may_nulls_set) then
+        NotNull
+      else
+        Top
+    | Some max_i, Some max_size when Z.geq max_i Z.zero ->
+      (* if maximum value in index interval < minimal size, return Null if all numbers in index interval are in must_nulls_set *)
+      if Z.lt max_i min_size && all_indexes_must_null min_i max_i then
+        Null
+        (* if maximum value in index interval < maximal size, return NotNull if no number in index interval is in may_nulls_set *)
+      else if Z.lt max_i max_size && not (MayNulls.exists (fun x -> Z.geq x min_i && Z.leq x max_i) may_nulls_set) then
+        NotNull
+      else
+        Top
+    (* if maximum number in interval is invalid, i.e. negative, return Top of value *)
+    | _ -> Top
+
+  (* helper functions *)
+  let must_nulls_remove i must_nulls_set min_size =
+    let rec compute_set acc i =
+      if Z.geq i min_size then 
+        acc
+      else
+        compute_set (MustNulls.add i acc) (Z.succ i) in
+    if MustNulls.is_bot must_nulls_set then
+      MustNulls.remove i (compute_set (MustNulls.empty ()) Z.zero)
+    else
+      MustNulls.remove i must_nulls_set
+  let must_nulls_filter cond must_nulls_set min_size = 
+    let rec compute_set acc i =
+      if Z.geq i min_size then 
+        acc
+      else
+        compute_set (MustNulls.add i acc) (Z.succ i) in
```

Comment:
This should be possible to do more concisely, e.g. `List.init` and then `MustNulls.from_list`